src/fn/csv-to-docs.js

```diff
@@ -0,0 +1,157 @@
+const trace = require('../lib/log').trace;
+const uuid5 = require('uuid/v5');
+const warn = require('../lib/log').warn;
+
+require('../lib/polyfill');
+
+module.exports = (projectDir, couchUrl) => {
+  if(!couchUrl) return skipFn('no couch URL set');
+  const couchUrlUuid = uuid5(couchUrl, uuid5.URL);
+
+  const csvDir = `${projectDir}/csv`;
+  if(!fs.exists(csvDir)) {
+    warn(`No csv directory found at ${csvDir}.`);
+    return Promise.resolve();
+  }
+
+  const jsonDir = `${projectDir}/json_docs`;
+  fs.mkdir(jsonDir);
+
+  const saveJsonDoc = doc => fs.writeJson(`${jsonDir}/${doc._id}.doc.json`, doc);
+
+  const model = {
+    docs: {},
+    references: [],
+  };
+  const addToModel = docs => {
+    docs.forEach(doc => {
+      model.docs[doc._id] = doc;
+      if(!model[doc.type]) model[doc.type] = [];
+      model[doc.type].push(doc);
+    });
+  };
+
+  return fs.recurseFiles(csvDir)
+    .filter(name => name.endsWith('.csv'))
+    .reduce((promiseChain, csv) =>
+        promiseChain
+          .then(() => {
+            info('Processing CSV file:', csv, '…');
+
+            const prefix = fs.path.basename(csv).split('.')[0];
+            switch(prefix) {
+              case 'person': return processPersons(csv);
+              case 'place':  return processPlaces(csv);
+              case 'report': return processReports(csv);
+              default: throw new Error(`Unrecognised CSV type ${prefix} for file ${csv}`);
+            }
+          })
+          .then(addToModel),
+      Promise.resolve())
+
+    .then(() => {
+      model.references.forEach(ref => {
+        ref.doc[ref.property] = model[ref.type][ref.index]._id;
+      });
+    })
+
+    .then(() => trace('Should now create all files for docs:', JSON.stringify(model, null, 2)))
+    .then(() => Promise.all(Object.values(model.docs).map(saveJsonDoc)));
+
+
+  function withId(json) {
+    const id = uuid(json);
+    json._id = id;
+    return json;
+  }
+
+  function loadCsv(csv) {
+    const raw = csvParse(fs.read(csv));
+    if(!raw.length) return { cols:[], rows:[] };
+    return {
+      cols: raw[0],
+      rows: raw.slice(1),
+    };
+  }
+
+  function processPersons(csv) {
+    return processContacts('person', csv);
+  }
+
+  function processPlaces(csv) {
+    const placeType = fs.path.basename(csv).split('.')[1];
+    return processContacts(placeType, csv);
+  }
+
+  function processReports(csv) {
+    return loadCsv(csv).rows
+      .map(r => console.log('TODO', r) || withId({
+        type: 'report',
+      }));
+  }
+
+  function processContacts(contactType, csv) {
+    const { rows, cols } = loadCsv(csv);
+    return rows
+      .map(r => {
+        const doc = { type:contactType };
+
+        function addReference(r) {
+          r.doc = doc;
+          model.references.push(r);
+        }
+
+        for(let i=0; i<cols.length; ++i) {
+          const { col, val, references } = parseColumn(cols[i], r[i]);
+          doc[col] = val;
```

Comment:
We'll need to handle nested fields at some point, eg: `fields.name:Alex` => `{ fields: { name: "Alex" } }`

Reply:
Is this done in contacts, or just reports?

Reply:
Probably just reports, but I can't say for sure.